ios/Peepz/Peepz/Model/PeepzModel.swift

```diff
@@ -0,0 +1,91 @@
+import Foundation
+import GoogleSignIn
+import Firebase
+
+class PeepzModel: NSObject, ObservableObject {
+    @Published var isAuthenticated = false
+    @Published var users = [GalleryItemViewState]()
+
+    var fireDatabaseRef = Database.database().reference()
+
+    override init() {
+        super.init()
+        configureGoogleSignIn()
+    }
+
+    private func configureGoogleSignIn() {
+        GIDSignIn.sharedInstance().clientID = FirebaseApp.app()?.options.clientID
+        GIDSignIn.sharedInstance().delegate = self
+    }
+}
+
+extension PeepzModel: GIDSignInDelegate {
+    func sign(_ signIn: GIDSignIn!, didSignInFor user: GIDGoogleUser!, withError error: Error?) {
+
+        if let error = error {
+            print("Sign In failed with error: \(error)")
+            return
+        }
+
+        guard let authentication = user.authentication else { return }
+        let credential = GoogleAuthProvider.credential(withIDToken: authentication.idToken,
+                                                       accessToken: authentication.accessToken)
+
+        firebaseAuth(credential: credential)
+    }
+
+    func sign(_ signIn: GIDSignIn!, didDisconnectWith user: GIDGoogleUser!, withError error: Error!) {
+        // Perform any operations when the user disconnects from app here.
+        print("Logged out: Yet to be implemented")
+    }
+
+    func firebaseAuth(credential: AuthCredential) {
+        Auth.auth().signIn(with: credential) { (authResult, error) in
+
+            if let error = error {
+                print("Firebase Login failed: \(error)")
+                return
+            }
+            self.isAuthenticated = true
+            self.getUsers() //ToDo: Should this be put in a different place? - yes
+        }
+    }
+
+    func signOut() {
+        do {
+            GIDSignIn.sharedInstance().signOut()
+            try Auth.auth().signOut()
+
+            isAuthenticated = false
+        } catch let signOutError as NSError {
+            print ("Error signing out: %@", signOutError)
+        }
+    }
+
+    func getUsers() {
+        let fifteenMinutes: Double = 60 * 15
+        var ref: DatabaseReference!
+
+        ref = Database.database().reference()
+        ref.child("wip/rooms/novoda/wall/").observe(.value, with: { [weak self] snapshot  in
+            // Get user value
+            guard let value = snapshot.value as? NSDictionary else {
+                print("Error: No users here")
+                return
+            }
+            self?.users = value
+                .allValues
+                .compactMap { User(dictionary: $0 as! NSDictionary) }
+                .sorted(by: { $0.lastSeen > $1.lastSeen })
+                .map({ user -> GalleryItemViewState in
```

Comment:
Maybe the contents of this map could be put into an extension of User? 

```
extension User {
    func galleryItemViewState() -> GalleryItemViewState  {

    return GalleryItemViewState(imageName: self.imageUrl,
                                                     location: self.location,
                                                     name: self.name,
                                                     isActive: self.isActive)
    }

    private func isActive() -> Bool {
        let fifteenMinutes: Double = 60 * 15
        let lastSeenInMinutes = self.lastSeen * 0.001
        return Date().timeIntervalSince1970 - lastSeenInMinutes < fifteenMinutes
    }
}
```

Reply:
or maybe have isActive as a computed var instead of a function

Reply:
Most of the time I put this logic into floating private function or into a dedicated Factory (if they are meant to be reused). The important bit is to have the domain object, the user in this case, as a parameter so it's easier to get the concatenation with Combine/RX: 

```
func toViewState(user: User) -> ViewState { ... }
```

```
userStream
    .map(toViewState) // <-- passing the function pointer will pass the user automagically for us 
```

I also like to put this function out of the scope of the ViewModel, in this way I'm 100% sure that the view state creation is stateless and that I'm always passing all the input in order to generate a ViewState

Reply:
Addressed in: 0417cc2